### PR TITLE
Toggle virtpoller beacon when toggling the virtualization entitlement

### PR DESCRIPTION
**Pull request by cbosdo**
## What does this PR change?

With the optimizations performed in the pillar generation, the
virtpoller beacon was no longer added. To enable or disable the
virtpoller beacon we need to:

 * generate a separate pillar file with the beacon configuration
 * refresh the pillar data on the minion

Since there could be other beacons setup in other pillar files, the
suma-minion.py pillar needs to deep merge each pillar data.

See [bsc#1172962](https://bugzilla.suse.com/show_bug.cgi?id=1172962)

## GUI diff

No difference.

- [X] **DONE**

## Documentation
- No documentation needed: no UI difference

- [X] **DONE**

## Test coverage
- No tests: hard to cover

- [X] **DONE**

## Links

Fixes SUSE/spacewalk#11731

- [X] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "ruby_rubocop"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

File: java/code/src/com/suse/manager/virtualization/VirtManagerSalt.java [hunk 1]
```diff
@@ -38,6 +44,8 @@
 public class VirtManagerSalt implements VirtManager {
```
Comment (chiaradiamarcelo) [inline java/code/src/com/suse/manager/virtualization/VirtManagerSalt.java:44]: Shouldn't this be called "SaltVirtManager" ?

File: java/code/src/com/suse/manager/virtualization/VirtManagerSalt.java [hunk 2]
```diff
@@ -166,6 +174,14 @@ public void updateLibvirtEngine(MinionServer minion) {
...
         saltApi.callSync(State.apply(Collections.singletonList("virt.engine-events"),
                 Optional.of(pillar)), minion.getMinionId());
+
+        if (minion.hasVirtualizationEntitlement()) {
+            minionVirtualizationPillarFileManager.generatePillarFile(minion);
+        }
+        else {
+            minionVirtualizationPillarFileManager.removePillarFile(minion.getMinionId());
+        }
+        SaltService.INSTANCE.refreshPillar(new MinionList(minion.getMinionId()));
```
Comment (chiaradiamarcelo) [inline java/code/src/com/suse/manager/virtualization/VirtManagerSalt.java:184]: why is this needed?

File: susemanager-utils/susemanager-sls/modules/pillar/suma_minion.py [hunk 3]
```diff
@@ -100,7 +100,10 @@ def ext_pillar(minion_id, *args):
         data_filename = os.path.join(MANAGER_PILLAR_DATA_PATH, minion_pillar_filename_prefix + suffix)
         if os.path.exists(data_filename):
             try:
-                ret.update(yaml.load(open(data_filename).read(), Loader=yaml.FullLoader))
+                ret = salt.utils.dictupdate.merge(
```
Comment (chiaradiamarcelo) [inline susemanager-utils/susemanager-sls/modules/pillar/suma_minion.py:103]: why is this needed here?

File: java/code/src/com/suse/manager/webui/services/pillar/MinionVirtualizationPillarGenerator.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+import org.apache.log4j.Logger;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Class for generating pillar data for the virtual hosts
+ */
+public class MinionVirtualizationPillarGenerator implements MinionPillarGenerator {
```
Comment (chiaradiamarcelo) [inline java/code/src/com/suse/manager/webui/services/pillar/MinionVirtualizationPillarGenerator.java:34]: We should probably make a test for this

**Comment by chiaradiamarcelo** [comment]
You probably should include the new pillar generator here:

https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/suse/manager/webui/services/pillar/MinionPillarManager.java#L37

File: java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java [hunk 5]
```diff
@@ -85,6 +95,10 @@ public void removeServerEntitlement(Server server, Entitlement ent) {
                 }
             });
         }
+
```
Comment (chiaradiamarcelo) [inline java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java:98]: while you are at it, maybe you can fix this part https://github.com/uyuni-project/uyuni/pull/2325/files?file-filters%5B%5D=.java#diff-8cc725b56315ba8fd87da99a53fb4fcdL76-R88 

The conditions here should be inverted, as in https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/redhat/rhn/manager/system/entitling/SystemEntitler.java#L128-L134

File: java/code/src/com/suse/manager/virtualization/VirtManagerSalt.java (hunk 1 above)
Comment (cbosdo) [reply java/code/src/com/suse/manager/virtualization/VirtManagerSalt.java:44]: Well, that class has been named like this by @lucidd during the refactoring.

File: java/code/src/com/suse/manager/virtualization/VirtManagerSalt.java (hunk 2 above)
Comment (cbosdo) [reply java/code/src/com/suse/manager/virtualization/VirtManagerSalt.java:184]: Without the `refreshPillar()` the pillar file is generated and the results can be seen in `salt 
pillar.items` but the change isn't reflected on the minion itself. Maybe there is a better place for this call, but without it the virtpoller is configured but not enabled on the minion

File: susemanager-utils/susemanager-sls/modules/pillar/suma_minion.py (hunk 3 above)
Comment (cbosdo) [reply susemanager-utils/susemanager-sls/modules/pillar/suma_minion.py:103]: There may be multiple beacons configured in different files and we don't want one to kill the other. Typically here I had:

```yaml
- beacons:
    pkgset:
      cookie: /var/cache/salt/minion/rpmdb.cookie
      interval: 5
```

That was kicked out by the virtpoller config

File: java/code/src/com/suse/manager/webui/services/pillar/MinionVirtualizationPillarGenerator.java (hunk 4 above)
Comment (cbosdo) [reply java/code/src/com/suse/manager/webui/services/pillar/MinionVirtualizationPillarGenerator.java:34]: sure, though that's not the most interesting test to have here. Covering the whole thing with a cucumber step would be better, though harder

File: java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java (hunk 5 above)
Comment (cbosdo) [reply java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java:98]: not sure I really understood where you want that change, but I'll try to find it

**Comment by cbosdo** [comment]
> You probably should include the new pillar generator here:
> 
> https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/suse/manager/webui/services/pillar/MinionPillarManager.java#L37

Thanks for the pointer. Will add it here.

File: java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java (hunk 5 above)
Comment (chiaradiamarcelo) [reply java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java:98]: https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java#L76-L78

This is wrong, and was introduced in https://github.com/uyuni-project/uyuni/commit/0d8fe90492c3bfd8620191a1d9c61af1b160f0f4

The right way is to do it as in https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/redhat/rhn/manager/system/entitling/SystemEntitler.java#L128-L134 , meaning check if the server is a minion, ,then update the groups pillar, and then the specific features like mointoring or virtualization

File: java/code/src/com/suse/manager/virtualization/VirtManagerSalt.java (hunk 1 above)
Comment (chiaradiamarcelo) [reply java/code/src/com/suse/manager/virtualization/VirtManagerSalt.java:44]: ok, maybe for a next PR then.

File: java/code/src/com/suse/manager/virtualization/VirtManagerSalt.java (hunk 2 above)
Comment (chiaradiamarcelo) [reply java/code/src/com/suse/manager/virtualization/VirtManagerSalt.java:184]: I see

File: java/code/src/com/suse/manager/virtualization/VirtManagerSalt.java (hunk 1 above)
Comment (renner) [reply java/code/src/com/suse/manager/virtualization/VirtManagerSalt.java:44]: I think we don't really have a written rule or a scheme for the naming of interfaces and implementations yet? For instance I recently created `MonitoringManager` (interface) and `FormulaMonitoringManager` (implementation) and did the naming the other way around. I would have been fine though to align with another scheme, so I guess we could maybe agree on a scheme and write it down?

File: java/code/src/com/suse/manager/virtualization/VirtManagerSalt.java (hunk 2 above)
Comment (renner) [reply java/code/src/com/suse/manager/virtualization/VirtManagerSalt.java:184]: As I understood all Salt calls should go through `SystemQuery` now,  so do we really want to make an exception for `refreshPillar()` and call it directly on the `SaltService` instance? Shouldn't we rather move `refreshPillar()` to the `SystemQuery` interface?

File: java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java (hunk 5 above)
Comment (renner) [reply java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java:98]: Thanks for pointing it out, and sorry as it was my mistake!

File: java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java [hunk 6]
```diff
@@ -40,8 +41,17 @@
...
     private MonitoringManager monitoringManager = new FormulaMonitoringManager();
+    private VirtManager virtManager;
+
+    /**
+     * @param virtManagerIn instance for managing virtual machines.
+     * @param monitoringManagerIn instance for handling monitoring configuration.
+     */
+    public SystemUnentitler(VirtManager virtManagerIn, MonitoringManager monitoringManagerIn) {
+        this.virtManager = virtManagerIn;
+        this.monitoringManager = monitoringManagerIn;
```
Comment (renner) [inline java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java:53]: Thanks for setting `monitoringManager` here in the constructor now, maybe we no longer need to initialize it above?

Comment (cbosdo) [reply java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java:53]: right

File: java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java (hunk 5 above)
Comment (chiaradiamarcelo) [reply java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java:96]: haha no worries!

File: java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java [hunk 7]
```diff
@@ -73,8 +82,8 @@ public void removeServerEntitlement(Server server, Entitlement ent) {
             unentitleServer(server, ent);
         }
 
-        if (EntitlementManager.MONITORING.equals(ent)) {
-            server.asMinionServer().ifPresent(s -> {
+        server.asMinionServer().ifPresent(s -> {
+            if (EntitlementManager.MONITORING.equals(ent)) {
                 ServerGroupManager.getInstance().updatePillarAfterGroupUpdateForServers(Arrays.asList(s));
```
Comment (chiaradiamarcelo) [inline java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java:87]: line 87 " ServerGroupManager.getInstance().updatePillarAfterGroupUpdateForServers(Arrays.asList(s)); "
should be executed despite the minion has a monitoring entitlement or not. So it should be outside this 'if' statement

File: java/code/src/com/suse/manager/virtualization/VirtManagerSalt.java (hunk 2 above)
Comment (cbosdo) [reply java/code/src/com/suse/manager/virtualization/VirtManagerSalt.java]: Done!

File: java/code/src/com/suse/manager/webui/services/pillar/MinionVirtualizationPillarGenerator.java (hunk 4 above)
Comment (cbosdo) [reply java/code/src/com/suse/manager/webui/services/pillar/MinionVirtualizationPillarGenerator.java:34]: Made that test and a cucumber one!

File: java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java (hunk 6 above)
Comment (cbosdo) [reply java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java:52]: done

File: java/code/src/com/suse/manager/webui/services/pillar/test/MinionVirtualizationPillarGeneratorTest.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+            map = new Yaml().loadAs(fi, Map.class);
+        }
+
+        assertTrue(map.containsKey("beacons"));
+        Map<String, Object> beacons = (Map<String, Object>) map.get("beacons");
+
+        assertTrue(beacons.containsKey("virtpoller"));
+        Map<String, Object> virtpoller = (Map<String, Object>)beacons.get("virtpoller");
+
+        virtpoller.containsKey("cache_file");
```
Comment (chiaradiamarcelo) [inline java/code/src/com/suse/manager/webui/services/pillar/test/MinionVirtualizationPillarGeneratorTest.java:70]: I guess you missed the "asserts" here

File: java/code/src/com/suse/manager/webui/services/pillar/test/MinionVirtualizationPillarGeneratorTest.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        virtpoller.containsKey("expire_time");
```
Comment (chiaradiamarcelo) [inline java/code/src/com/suse/manager/webui/services/pillar/test/MinionVirtualizationPillarGeneratorTest.java:71]: same

File: java/code/src/com/suse/manager/webui/services/pillar/test/MinionVirtualizationPillarGeneratorTest.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        virtpoller.containsKey("interval");
```
Comment (chiaradiamarcelo) [inline java/code/src/com/suse/manager/webui/services/pillar/test/MinionVirtualizationPillarGeneratorTest.java:72]: same

Comment (cbosdo) [reply java/code/src/com/suse/manager/webui/services/pillar/test/MinionVirtualizationPillarGeneratorTest.java]: oops! Fixed now

File: java/code/src/com/suse/manager/webui/services/pillar/test/MinionVirtualizationPillarGeneratorTest.java (hunk 9 above)
Comment (cbosdo) [reply java/code/src/com/suse/manager/webui/services/pillar/test/MinionVirtualizationPillarGeneratorTest.java]: fixed too

File: java/code/src/com/suse/manager/webui/services/pillar/test/MinionVirtualizationPillarGeneratorTest.java (hunk 8 above)
Comment (cbosdo) [reply java/code/src/com/suse/manager/webui/services/pillar/test/MinionVirtualizationPillarGeneratorTest.java]: Fixed

File: java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java (hunk 7 above)
Comment (cbosdo) [reply java/code/src/com/redhat/rhn/manager/system/entitling/SystemUnentitler.java]: Fixed

File: java/code/src/com/suse/manager/webui/services/pillar/test/MinionVirtualizationPillarGeneratorTest.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+        assertTrue(map.containsKey("beacons"));
+        Map<String, Object> beacons = (Map<String, Object>) map.get("beacons");
+
+        assertTrue(beacons.containsKey("virtpoller"));
+        Map<String, Object> virtpoller = (Map<String, Object>)beacons.get("virtpoller");
+
+        assertTrue(virtpoller.containsKey("cache_file"));
+        assertTrue(virtpoller.containsKey("expire_time"));
+        assertTrue(virtpoller.containsKey("interval"));
```
Comment (chiaradiamarcelo) [inline java/code/src/com/suse/manager/webui/services/pillar/test/MinionVirtualizationPillarGeneratorTest.java:72]: I guess you missed the asserts here

Comment (cbosdo) [reply java/code/src/com/suse/manager/webui/services/pillar/test/MinionVirtualizationPillarGeneratorTest.java:72]: Yes, but it's fixed now, isn't it?
